src/game.cpp

```diff
@@ -10715,7 +10717,9 @@ bool game::unload( item &it )
             u.moves -= mv;
             return true;
         } ), it.contents.end() );
-        it.on_contents_changed();
+        if( it.contents.size() != old_size || it.contents.front().charges != old_charges ) {
```

Comment:
This looks a bit fragile and won't work as expected if an item from somewhere in the middle of `it.contents` changes its charge.  I suppose something like this would be a bit better (though, still quite ugly):

```c++
bool changed = false;
it.contents.erase( std::remove_if( it.contents.begin(), it.contents.end(), [this, &changed]( item& e ) {
    long old_charges = e.charges;
    const bool consumed = add_or_drop_with_msg( u, e );
    changed = changed || consumed || e.charges != old_charges;
    if( consumed ) {
        u.mod_moves( -u.item_handling_cost( e ) );
    }
    return consumed;
} ), it.contents.end() );

if( changed ) {
    it.on_contents_changed();
}
```